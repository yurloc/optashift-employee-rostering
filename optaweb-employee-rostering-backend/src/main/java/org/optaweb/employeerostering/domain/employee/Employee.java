/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.optaweb.employeerostering.domain.employee;

import java.util.Collection;
import java.util.Set;

import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.JoinColumn;
import javax.persistence.JoinTable;
import javax.persistence.ManyToMany;
import javax.persistence.ManyToOne;
import javax.persistence.Table;
import javax.persistence.UniqueConstraint;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Pattern;
import javax.validation.constraints.Size;

import org.optaweb.employeerostering.domain.common.AbstractPersistable;
import org.optaweb.employeerostering.domain.contract.Contract;
import org.optaweb.employeerostering.domain.skill.Skill;

@Entity
@Table(uniqueConstraints = @UniqueConstraint(columnNames = {"tenantId", "name"}))
public class Employee extends AbstractPersistable {

    @NotNull
    @Size(min = 1, max = 120)
    @Pattern(regexp = "^(?!\\s).*(?<!\\s)$", message = "Name should not contain any leading or trailing whitespaces")
    private String name;

    @NotNull
    @ManyToOne
    private Contract contract;

    @NotNull
    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(name = "EmployeeSkillProficiencySet",
            joinColumns = @JoinColumn(name = "employeeId", referencedColumnName = "id"),
            inverseJoinColumns = @JoinColumn(name = "skillId", referencedColumnName = "id")
    )
    private Set<Skill> skillProficiencySet;

    @SuppressWarnings("unused")
    public Employee() {
    }

    public Employee(Integer tenantId, String name, Contract contract, Set<Skill> skillProficiencySet) {
        super(tenantId);
        this.name = name;
        this.contract = contract;
        this.skillProficiencySet = skillProficiencySet;
    }

    public boolean hasSkill(Skill skill) {
        return skillProficiencySet.contains(skill);
    }

    public boolean hasSkills(Collection<Skill> skills) {
        return skillProficiencySet.containsAll(skills);
    }

    @Override
    public String toString() {
        return name;
    }

    // ************************************************************************
    // Simple getters and setters
    // ************************************************************************

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Set<Skill> getSkillProficiencySet() {
        return skillProficiencySet;
    }

    public void setSkillProficiencySet(Set<Skill> skillProficiencySet) {
        this.skillProficiencySet = skillProficiencySet;
    }

    public Contract getContract() {
        return contract;
    }

    public void setContract(Contract contract) {
        this.contract = contract;
    }

    @Override
    public boolean equals(Object o) {
        if (o instanceof Employee) {
            Employee other = (Employee) o;
            return this.name.equals(other.getName()) &&
                    this.contract.equals(other.getContract()) &&
                    this.skillProficiencySet.equals(other.getSkillProficiencySet());
        } else {
            return false;
        }
    }

    @Override
    public int hashCode() {
        return 31 * ((31 * name.hashCode()) ^ contract.hashCode()) ^ skillProficiencySet.hashCode();
    }
}
