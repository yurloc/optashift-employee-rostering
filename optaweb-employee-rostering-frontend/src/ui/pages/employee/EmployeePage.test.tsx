/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import TypeaheadSelectInput from 'ui/components/TypeaheadSelectInput';
import MultiTypeaheadSelectInput from 'ui/components/MultiTypeaheadSelectInput';
import { Sorter } from 'types';
import { Employee } from 'domain/Employee';
import { act } from 'react-dom/test-utils';
import { useTranslation } from 'react-i18next';
import { getRouterProps } from 'util/BookmarkableTestUtils';
import { EmployeesPage, Props } from './EmployeesPage';

describe('Employees page', () => {
  it('should render correctly with no employees', () => {
    const employeesPage = shallow(<EmployeesPage {...noEmployees} />);
    expect(toJson(employeesPage)).toMatchSnapshot();
  });

  it('should render correctly with a few employees', () => {
    const employeesPage = shallow(<EmployeesPage {...twoEmployees} />);
    expect(toJson(employeesPage)).toMatchSnapshot();
  });

  it('should render the viewer correctly', () => {
    const employeesPage = new EmployeesPage(twoEmployees);
    const spot = twoEmployees.tableData[1];
    const viewer = shallow(employeesPage.renderViewer(spot));
    expect(toJson(viewer)).toMatchSnapshot();
  });

  it('should render the editor correctly', () => {
    const employeesPage = new EmployeesPage(twoEmployees);
    const spot = twoEmployees.tableData[1];
    const editor = shallow(employeesPage.renderEditor(spot));
    expect(toJson(editor)).toMatchSnapshot();
  });

  it('should update properties on change', () => {
    const employeesPage = new EmployeesPage(twoEmployees);
    const setProperty = jest.fn();
    const editor = employeesPage.editDataRow(employeesPage.getInitialStateForNewRow(), setProperty);
    const nameCol = shallow(editor[0] as React.ReactElement);
    nameCol.simulate('change', { currentTarget: { value: 'Test' } });
    expect(setProperty).toBeCalled();
    expect(setProperty).toBeCalledWith('name', 'Test');

    setProperty.mockClear();
    const contractCol = mount(editor[1] as React.ReactElement);
    act(() => {
      contractCol.find(TypeaheadSelectInput).props().onChange(twoEmployees.contractList[0]);
    });
    expect(setProperty).toBeCalled();
    expect(setProperty).toBeCalledWith('contract', twoEmployees.contractList[0]);

    setProperty.mockClear();
    const skillProficiencySetCol = mount(editor[2] as React.ReactElement);
    act(() => {
      skillProficiencySetCol.find(MultiTypeaheadSelectInput).props().onChange([twoEmployees.skillList[0]]);
    });
    expect(setProperty).toBeCalled();
    expect(setProperty).toBeCalledWith('skillProficiencySet', [twoEmployees.skillList[0]]);
  });

  it('should call addEmployee on addData', () => {
    const employeesPage = new EmployeesPage(twoEmployees);
    const employee: Employee = {
      name: 'Employee',
      skillProficiencySet: [],
      contract: twoEmployees.contractList[0],
      tenantId: 0,
      id: 1,
      version: 0,
    };
    employeesPage.addData(employee);
    expect(twoEmployees.addEmployee).toBeCalled();
    expect(twoEmployees.addEmployee).toBeCalledWith(employee);
  });

  it('should call updateEmployee on updateData', () => {
    const employeesPage = new EmployeesPage(twoEmployees);
    const employee: Employee = {
      name: 'Employee',
      skillProficiencySet: [],
      contract: twoEmployees.contractList[0],
      tenantId: 0,
      id: 1,
      version: 0,
    };
    employeesPage.updateData(employee);
    expect(twoEmployees.updateEmployee).toBeCalled();
    expect(twoEmployees.updateEmployee).toBeCalledWith(employee);
  });

  it('should call removeEmployee on removeData', () => {
    const employeesPage = new EmployeesPage(twoEmployees);
    const employee: Employee = {
      name: 'Employee',
      skillProficiencySet: [],
      contract: twoEmployees.contractList[0],
      tenantId: 0,
      id: 1,
      version: 0,
    };
    employeesPage.removeData(employee);
    expect(twoEmployees.removeEmployee).toBeCalled();
    expect(twoEmployees.removeEmployee).toBeCalledWith(employee);
  });

  it('should return a filter that match by name, skills and contract', () => {
    const employeesPage = new EmployeesPage(twoEmployees);
    const filter = employeesPage.getFilter();

    expect(twoEmployees.tableData.filter(filter('1'))).toEqual([twoEmployees.tableData[0], twoEmployees.tableData[1]]);
    expect(twoEmployees.tableData.filter(filter('Skill 1'))).toEqual([twoEmployees.tableData[1]]);
    expect(twoEmployees.tableData.filter(filter('2'))).toEqual([twoEmployees.tableData[1]]);
    expect(twoEmployees.tableData.filter(filter('Contract 2'))).toEqual([twoEmployees.tableData[1]]);
    expect(twoEmployees.tableData.filter(filter('Employee 2'))).toEqual([twoEmployees.tableData[1]]);
  });

  it('should return a sorter that sort by name and contract', () => {
    const employeesPage = new EmployeesPage(twoEmployees);
    const nameSorter = employeesPage.getSorters()[0] as Sorter<Employee>;
    let list = [twoEmployees.tableData[1], twoEmployees.tableData[0]];
    expect(list.sort(nameSorter)).toEqual(twoEmployees.tableData);
    list = [twoEmployees.tableData[1], twoEmployees.tableData[0]];
    const contractSorter = employeesPage.getSorters()[1] as Sorter<Employee>;
    expect(list.sort(contractSorter)).toEqual(twoEmployees.tableData);
    expect(employeesPage.getSorters()[2]).toBeNull();
  });

  it('should go to the Contract page if the user click on the link', () => {
    const employeesPage = shallow(<EmployeesPage
      {...noEmployees}
    />);
    shallow((employeesPage.find('Trans').prop('components') as any)[2]).simulate('click');
    expect(noEmployees.history.push).toBeCalled();
    expect(noEmployees.history.push).toBeCalledWith('/0/contracts');
  });

  it('should treat incompleted data as incomplete', () => {
    const employeesPage = new EmployeesPage(twoEmployees);
    const noName = {
      tenantId: 0,
      skillProficiencySet: [],
      contract: twoEmployees.contractList[0],
    };
    const result1 = employeesPage.isDataComplete(noName);
    expect(result1).toEqual(false);

    const noSkills = {
      tenantId: 0,
      name: 'Name',
      contract: twoEmployees.contractList[0],
    };
    const result2 = employeesPage.isDataComplete(noSkills);
    expect(result2).toEqual(false);

    const noContract = {
      tenantId: 0,
      name: 'Name',
      skillProficiencySet: [],
    };
    const result3 = employeesPage.isDataComplete(noContract);
    expect(result3).toEqual(false);

    const completed = {
      tenantId: 0,
      name: 'Name',
      skillProficiencySet: [],
      contract: twoEmployees.contractList[0],
    };
    const result4 = employeesPage.isDataComplete(completed);
    expect(result4).toEqual(true);
  });

  it('should treat empty name as invalid', () => {
    const employeesPage = new EmployeesPage(twoEmployees);
    const noName = {
      tenantId: 0,
      name: '',
      skillProficiencySet: [],
      contract: twoEmployees.contractList[0],
    };
    const result1 = employeesPage.isValid(noName);
    expect(result1).toEqual(false);
  });

  it('should treat non-empty name as valid', () => {
    const employeesPage = new EmployeesPage(twoEmployees);
    const components = twoEmployees.tableData[0];
    const result = employeesPage.isValid(components);
    expect(result).toEqual(true);
  });
});

const noEmployees: Props = {
  ...useTranslation('EmployeePage'),
  tReady: true,
  tenantId: 0,
  title: 'Employees',
  columnTitles: ['Name', 'Contract', 'Skill Set'],
  tableData: [],
  skillList: [],
  contractList: [],
  addEmployee: jest.fn(),
  updateEmployee: jest.fn(),
  removeEmployee: jest.fn(),
  ...getRouterProps('/contacts', {}),
};

const twoEmployees: Props = {
  ...useTranslation('EmployeePage'),
  tReady: true,
  tenantId: 0,
  title: 'Employees',
  columnTitles: ['Name', 'Contract', 'Skill Set'],
  tableData: [{
    id: 0,
    version: 0,
    tenantId: 0,
    name: 'Employee 1',
    skillProficiencySet: [],
    contract: {
      tenantId: 0,
      id: 0,
      version: 0,
      name: 'Contract 1',
      maximumMinutesPerDay: null,
      maximumMinutesPerWeek: null,
      maximumMinutesPerMonth: null,
      maximumMinutesPerYear: null,
    },
  },
  {
    id: 1,
    version: 0,
    tenantId: 0,
    name: 'Employee 2',
    skillProficiencySet: [{ tenantId: 0, name: 'Skill 1' }, { tenantId: 0, name: 'Skill 2' }],
    contract: {
      tenantId: 0,
      id: 1,
      version: 0,
      name: 'Contract 2',
      maximumMinutesPerDay: null,
      maximumMinutesPerWeek: null,
      maximumMinutesPerMonth: null,
      maximumMinutesPerYear: null,
    },
  }],
  skillList: [{ tenantId: 0, name: 'Skill 1' }, { tenantId: 0, name: 'Skill 2' }],
  contractList: [
    {
      tenantId: 0,
      id: 0,
      version: 0,
      name: 'Contract 1',
      maximumMinutesPerDay: null,
      maximumMinutesPerWeek: null,
      maximumMinutesPerMonth: null,
      maximumMinutesPerYear: null,
    },
    {
      tenantId: 0,
      id: 1,
      version: 0,
      name: 'Contract 2',
      maximumMinutesPerDay: null,
      maximumMinutesPerWeek: null,
      maximumMinutesPerMonth: null,
      maximumMinutesPerYear: null,
    },
  ],
  addEmployee: jest.fn(),
  updateEmployee: jest.fn(),
  removeEmployee: jest.fn(),
  ...getRouterProps('/employees', {}),
};
