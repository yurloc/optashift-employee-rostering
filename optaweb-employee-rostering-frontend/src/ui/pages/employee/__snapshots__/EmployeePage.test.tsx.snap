// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Employees page should render correctly with a few employees 1`] = `
<Fragment>
  <Level
    gutter="sm"
    style={
      Object {
        "backgroundColor": "var(--pf-global--BackgroundColor--200)",
        "padding": "5px 5px 5px 5px",
      }
    }
  >
    <LevelItem>
      <FilterComponent
        filterText=""
        onChange={[Function]}
      />
    </LevelItem>
    <LevelItem
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <Button
        isDisabled={false}
        onClick={[Function]}
      >
        Trans(i18nKey=add)
      </Button>
      <Pagination
        itemCount={2}
        onPerPageSelect={[Function]}
        onSetPage={[Function]}
        page={1}
        perPage={10}
        widgetId="pagination-options-menu-top"
      />
    </LevelItem>
  </Level>
  <Table
    borders={true}
    caption="Employees"
    cells={
      Array [
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Name",
          "transforms": Array [
            [Function],
          ],
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Contract",
          "transforms": Array [
            [Function],
          ],
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Skill Set",
          "transforms": undefined,
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "",
          "transforms": undefined,
        },
      ]
    }
    className=""
    contentId="expanded-content"
    dropdownDirection="down"
    dropdownPosition="right"
    expandId="expandable-toggle"
    gridBreakPoint="grid-md"
    onSort={[Function]}
    role="grid"
    rowLabeledBy="simple-node"
    rows={
      Array [
        Object {
          "cells": Array [
            Object {
              "title": <EditableComponent
                editor={
                  <TextInput
                    aria-label="Name"
                    className=""
                    defaultValue="Employee 1"
                    isDisabled={false}
                    isReadOnly={false}
                    isRequired={false}
                    isValid={true}
                    name="name"
                    onChange={[Function]}
                    type="text"
                  />
                }
                isEditing={false}
                viewer={
                  <Text>
                    Employee 1
                  </Text>
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <StatefulTypeaheadSelectInput
                    emptyText="Trans(i18nKey=selectAContract)"
                    onChange={[Function]}
                    optionToStringMap={[Function]}
                    options={
                      Array [
                        Object {
                          "id": 0,
                          "maximumMinutesPerDay": null,
                          "maximumMinutesPerMonth": null,
                          "maximumMinutesPerWeek": null,
                          "maximumMinutesPerYear": null,
                          "name": "Contract 1",
                          "tenantId": 0,
                          "version": 0,
                        },
                        Object {
                          "id": 1,
                          "maximumMinutesPerDay": null,
                          "maximumMinutesPerMonth": null,
                          "maximumMinutesPerWeek": null,
                          "maximumMinutesPerYear": null,
                          "name": "Contract 2",
                          "tenantId": 0,
                          "version": 0,
                        },
                      ]
                    }
                    value={
                      Object {
                        "id": 0,
                        "maximumMinutesPerDay": null,
                        "maximumMinutesPerMonth": null,
                        "maximumMinutesPerWeek": null,
                        "maximumMinutesPerYear": null,
                        "name": "Contract 1",
                        "tenantId": 0,
                        "version": 0,
                      }
                    }
                  />
                }
                isEditing={false}
                viewer={
                  <Text>
                    Contract 1
                  </Text>
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <StatefulMultiTypeaheadSelectInput
                    emptyText="Trans(i18nKey=selectSkillProficiencies)"
                    onChange={[Function]}
                    optionToStringMap={[Function]}
                    options={
                      Array [
                        Object {
                          "name": "Skill 1",
                          "tenantId": 0,
                        },
                        Object {
                          "name": "Skill 2",
                          "tenantId": 0,
                        },
                      ]
                    }
                    value={Array []}
                  />
                }
                isEditing={false}
                viewer={
                  <ChipGroup
                    className=""
                    collapsedText="\${remaining} more"
                    expandedText="Show Less"
                    withToolbar={false}
                  />
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <span
                    style={
                      Object {
                        "display": "grid",
                        "gridColumnGap": "5px",
                        "gridTemplateColumns": "1fr auto auto",
                      }
                    }
                  >
                    <span />
                    <Button
                      aria-label="Save"
                      onClick={[Function]}
                      variant="link"
                    >
                      <SaveIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                        title={null}
                      />
                    </Button>
                    <Button
                      aria-label="Cancel"
                      onClick={[Function]}
                      variant="link"
                    >
                      <CloseIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                        title={null}
                      />
                    </Button>
                  </span>
                }
                isEditing={false}
                viewer={
                  <span
                    style={
                      Object {
                        "display": "grid",
                        "gridColumnGap": "5px",
                        "gridTemplateColumns": "1fr auto auto",
                      }
                    }
                  >
                    <span />
                    <Button
                      aria-label="Edit"
                      onClick={[Function]}
                      variant="link"
                    >
                      <EditIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                        title={null}
                      />
                    </Button>
                    <Button
                      aria-label="Delete"
                      onClick={[Function]}
                      variant="link"
                    >
                      <TrashIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                        title={null}
                      />
                    </Button>
                  </span>
                }
              />,
            },
          ],
        },
        Object {
          "cells": Array [
            Object {
              "title": <EditableComponent
                editor={
                  <TextInput
                    aria-label="Name"
                    className=""
                    defaultValue="Employee 2"
                    isDisabled={false}
                    isReadOnly={false}
                    isRequired={false}
                    isValid={true}
                    name="name"
                    onChange={[Function]}
                    type="text"
                  />
                }
                isEditing={false}
                viewer={
                  <Text>
                    Employee 2
                  </Text>
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <StatefulTypeaheadSelectInput
                    emptyText="Trans(i18nKey=selectAContract)"
                    onChange={[Function]}
                    optionToStringMap={[Function]}
                    options={
                      Array [
                        Object {
                          "id": 0,
                          "maximumMinutesPerDay": null,
                          "maximumMinutesPerMonth": null,
                          "maximumMinutesPerWeek": null,
                          "maximumMinutesPerYear": null,
                          "name": "Contract 1",
                          "tenantId": 0,
                          "version": 0,
                        },
                        Object {
                          "id": 1,
                          "maximumMinutesPerDay": null,
                          "maximumMinutesPerMonth": null,
                          "maximumMinutesPerWeek": null,
                          "maximumMinutesPerYear": null,
                          "name": "Contract 2",
                          "tenantId": 0,
                          "version": 0,
                        },
                      ]
                    }
                    value={
                      Object {
                        "id": 1,
                        "maximumMinutesPerDay": null,
                        "maximumMinutesPerMonth": null,
                        "maximumMinutesPerWeek": null,
                        "maximumMinutesPerYear": null,
                        "name": "Contract 2",
                        "tenantId": 0,
                        "version": 0,
                      }
                    }
                  />
                }
                isEditing={false}
                viewer={
                  <Text>
                    Contract 2
                  </Text>
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <StatefulMultiTypeaheadSelectInput
                    emptyText="Trans(i18nKey=selectSkillProficiencies)"
                    onChange={[Function]}
                    optionToStringMap={[Function]}
                    options={
                      Array [
                        Object {
                          "name": "Skill 1",
                          "tenantId": 0,
                        },
                        Object {
                          "name": "Skill 2",
                          "tenantId": 0,
                        },
                      ]
                    }
                    value={
                      Array [
                        Object {
                          "name": "Skill 1",
                          "tenantId": 0,
                        },
                        Object {
                          "name": "Skill 2",
                          "tenantId": 0,
                        },
                      ]
                    }
                  />
                }
                isEditing={false}
                viewer={
                  <ChipGroup
                    className=""
                    collapsedText="\${remaining} more"
                    expandedText="Show Less"
                    withToolbar={false}
                  >
                    <Chip
                      className=""
                      closeBtnAriaLabel="close"
                      component="div"
                      isOverflowChip={false}
                      isReadOnly={true}
                      onClick={[Function]}
                      tooltipPosition="top"
                    >
                      Skill 1
                    </Chip>
                    <Chip
                      className=""
                      closeBtnAriaLabel="close"
                      component="div"
                      isOverflowChip={false}
                      isReadOnly={true}
                      onClick={[Function]}
                      tooltipPosition="top"
                    >
                      Skill 2
                    </Chip>
                  </ChipGroup>
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <span
                    style={
                      Object {
                        "display": "grid",
                        "gridColumnGap": "5px",
                        "gridTemplateColumns": "1fr auto auto",
                      }
                    }
                  >
                    <span />
                    <Button
                      aria-label="Save"
                      onClick={[Function]}
                      variant="link"
                    >
                      <SaveIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                        title={null}
                      />
                    </Button>
                    <Button
                      aria-label="Cancel"
                      onClick={[Function]}
                      variant="link"
                    >
                      <CloseIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                        title={null}
                      />
                    </Button>
                  </span>
                }
                isEditing={false}
                viewer={
                  <span
                    style={
                      Object {
                        "display": "grid",
                        "gridColumnGap": "5px",
                        "gridTemplateColumns": "1fr auto auto",
                      }
                    }
                  >
                    <span />
                    <Button
                      aria-label="Edit"
                      onClick={[Function]}
                      variant="link"
                    >
                      <EditIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                        title={null}
                      />
                    </Button>
                    <Button
                      aria-label="Delete"
                      onClick={[Function]}
                      variant="link"
                    >
                      <TrashIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                        title={null}
                      />
                    </Button>
                  </span>
                }
              />,
            },
          ],
        },
      ]
    }
    sortBy={Object {}}
    variant={null}
  >
    <TableHeader />
    <TableBody />
  </Table>
</Fragment>
`;

exports[`Employees page should render correctly with no employees 1`] = `
<EmptyState
  variant="full"
>
  <EmptyStateIcon
    icon={[Function]}
  />
  <Trans
    components={
      Array [
        <Title
          headingLevel="h5"
          size="lg"
        />,
        <EmptyStateBody />,
        <Button
          aria-label="Contracts Page"
          onClick={[Function]}
          variant="primary"
        />,
      ]
    }
    i18nKey="noContracts"
    t={[MockFunction]}
  />
</EmptyState>
`;

exports[`Employees page should render the editor correctly 1`] = `
<tr>
  <td
    key="0"
  >
    <TextInput
      aria-label="Name"
      className=""
      defaultValue="Employee 2"
      isDisabled={false}
      isReadOnly={false}
      isRequired={false}
      isValid={true}
      key="0"
      name="name"
      onChange={[Function]}
      type="text"
    />
  </td>
  <td
    key="1"
  >
    <StatefulTypeaheadSelectInput
      emptyText="Trans(i18nKey=selectAContract)"
      key="1"
      onChange={[Function]}
      optionToStringMap={[Function]}
      options={
        Array [
          Object {
            "id": 0,
            "maximumMinutesPerDay": null,
            "maximumMinutesPerMonth": null,
            "maximumMinutesPerWeek": null,
            "maximumMinutesPerYear": null,
            "name": "Contract 1",
            "tenantId": 0,
            "version": 0,
          },
          Object {
            "id": 1,
            "maximumMinutesPerDay": null,
            "maximumMinutesPerMonth": null,
            "maximumMinutesPerWeek": null,
            "maximumMinutesPerYear": null,
            "name": "Contract 2",
            "tenantId": 0,
            "version": 0,
          },
        ]
      }
      value={
        Object {
          "id": 1,
          "maximumMinutesPerDay": null,
          "maximumMinutesPerMonth": null,
          "maximumMinutesPerWeek": null,
          "maximumMinutesPerYear": null,
          "name": "Contract 2",
          "tenantId": 0,
          "version": 0,
        }
      }
    />
  </td>
  <td
    key="2"
  >
    <StatefulMultiTypeaheadSelectInput
      emptyText="Trans(i18nKey=selectSkillProficiencies)"
      key="2"
      onChange={[Function]}
      optionToStringMap={[Function]}
      options={
        Array [
          Object {
            "name": "Skill 1",
            "tenantId": 0,
          },
          Object {
            "name": "Skill 2",
            "tenantId": 0,
          },
        ]
      }
      value={
        Array [
          Object {
            "name": "Skill 1",
            "tenantId": 0,
          },
          Object {
            "name": "Skill 2",
            "tenantId": 0,
          },
        ]
      }
    />
  </td>
</tr>
`;

exports[`Employees page should render the viewer correctly 1`] = `
<tr>
  <td
    key="0"
  >
    <Text
      key="0"
    >
      Employee 2
    </Text>
  </td>
  <td
    key="1"
  >
    <Text
      key="1"
    >
      Contract 2
    </Text>
  </td>
  <td
    key="2"
  >
    <ChipGroup
      className=""
      collapsedText="\${remaining} more"
      expandedText="Show Less"
      key="2"
      withToolbar={false}
    >
      <Chip
        className=""
        closeBtnAriaLabel="close"
        component="div"
        isOverflowChip={false}
        isReadOnly={true}
        key="Skill 1"
        onClick={[Function]}
        tooltipPosition="top"
      >
        Skill 1
      </Chip>
      <Chip
        className=""
        closeBtnAriaLabel="close"
        component="div"
        isOverflowChip={false}
        isReadOnly={true}
        key="Skill 2"
        onClick={[Function]}
        tooltipPosition="top"
      >
        Skill 2
      </Chip>
    </ChipGroup>
  </td>
</tr>
`;
